src/kOS/Serialization/ConfigNodeFormatter.cs

```diff
@@ -0,0 +1,184 @@
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = ListKey;
+
+            var indexedList = list.Select((value, index) => new { Value = value, Index = index.ToString() });
+
+            foreach (var item in indexedList) {
+                HandleValue(innerNode, item.Index, item.Value);
+            }
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ScalarToConfigNode(string name, string type, string value)
+        {
+            ConfigNode configNode = new ConfigNode();
+            configNode.name = name;
+
+            // ConfigNode stores all values as strings so we have to store non-string values as nodes, otherwise
+            // the type information would be lost
+            ConfigNode innerNode = new ConfigNode();
+            innerNode.name = type;
+
+            innerNode.AddValue(ValueKey, value);
+
+            configNode.AddNode(innerNode);
+
+            return configNode;
+        }
+
+        private ConfigNode ToConfigNode(string name, Dump dump)
+        {
+            ConfigNode configNode = new ConfigNode();
+
+            configNode.name = name;
+
+            foreach (var entry in dump) {
+                HandleValue(configNode, entry.Key.ToString(), entry.Value);
+            }
+
+            return configNode;
+        }
+
+        private void HandleValue(ConfigNode configNode, string key, object value)
+        {
+            if (value is Dump)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as Dump));
+            } else if (value is List<object>)
+            {
+                configNode.AddNode(key, ToConfigNode(key, value as List<object>));
+            } else if (value is int) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, Convert.ToString(value)));
+            } else if (value is double) {
+                configNode.AddNode(key, ScalarToConfigNode(key, ScalarKey, ((double)value).ToString("R")));
+
+            } else if (value is bool) {
+                configNode.AddNode(key, ScalarToConfigNode(key, BooleanKey, Convert.ToString(value)));
+            } else
+            {
+                configNode.AddValue(key, value);
+            }
+        }
+
+        private List<object> ListFromConfigNode(ConfigNode configNode)
+        {
+            List<object> result = new List<object>();
+
+            int i = 0;
+
+            while (true) {
+                if (configNode.HasValue(i.ToString()))
+                {
+                    result.Add(configNode.GetValue(i.ToString()));
+                } else if (configNode.HasNode(i.ToString()))
+                {
+                    result.Add(ObjectFromConfigNode(configNode.GetNode(i.ToString())));
+                } else
+                {
+                    break;
+                }
+
+                i++;
+            }
```

Comment:
I'm not a fan of writing a finite loop by using what looks like an infinite loop in the header (``while (true)``) but with an explicit conditional break, unless there's a really complex set of circumstances causing that break that would greatly uglify the loop header, or it's something like an event dispatcher or a daemon that's expected to stay alive a long time.

This doesn't look like one of those cases.  It just looks like a finite loop where it consumes input until input is exhausted, and it's just that you can't see ahead of time how many iterations that will take.

I'd prefer something like:
```
bool moreValues = true;
while (moreValues)
{
    if .....
    else if ......
    else
        moreValues = false;
}
```
It's generally easier for people reading it to follow the logic that way.